File: Difficulty: Medium/Count Occurences of Anagrams/count-occurences-of-anagrams.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
// User function template for C++
class Solution {
  public:
    int search(string &pat, string &txt) {
        // code here
         map<char, int> mp;
    for (int i = 0; i < pat.size(); i++)
        mp[pat[i]]++;
    int count = mp.size();
    int i = 0, j = 0;
    int N = txt.size();
    int ans = 0;
    while (j < N)
    {
        if (mp.find(txt[j]) != mp.end())
        {
            mp[txt[j]]--;
            if (mp[txt[j]] == 0)
                count--;
        }
        if (j - i + 1 < pat.size())
            j++;
        else
        {
            if (count == 0)
                ans++;
            if (mp.find(txt[i]) != mp.end())
            {
                mp[txt[i]]++;
                if (mp[txt[i]] == 1)
                    count++;
            }
            i++;
            j++;
        }
    }
    return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string pat, txt;
        cin >> txt >> pat;
        Solution ob;
        auto ans = ob.search(pat, txt);
        cout << ans << "\n";

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Count Occurences of Anagrams/count-occurences-of-anagrams.java
// User function Template for Java

class Solution {

  static int search(String pat, String txt) {
        int i = 0;
        int j = 0;
        HashMap<Character, Integer> mp1 = new HashMap<>();
        HashMap<Character, Integer> mp2 = new HashMap<>();
        for (int k = 0; k < pat.length(); k++) {
            char x = pat.charAt(k);
            mp1.put(x, mp1.getOrDefault(x, 0) + 1);
        }
        int window = pat.length();
        int count = 0;
        while (j < txt.length()) {
            char charJ = txt.charAt(j);
            mp2.put(charJ, mp2.getOrDefault(charJ, 0) + 1);
            if (j - i + 1 < window) {

            } else if (j - i + 1 == window) {
                if (mp1.equals(mp2))
                    count++;
                char charI = txt.charAt(i);
                mp2.put(charI, mp2.get(charI) - 1);
                if (mp2.get(charI) == 0)
                    mp2.remove(charI);
                i++;
            }
            j++;
        }
        return count;

    }
}